### PR TITLE
[gtk] protect against nulls

### LINKED ISSUE #7592: Error opening darktable.
Hi,
The program opens but then closes showing the following error.
image

I am also attaching the back trace:
[darktable_bt_IHOAW0.txt](https://github.com/darktable-org/darktable/files/5758645/darktable_bt_IHOAW0.txt)

Not sure what is wrong. This happened after I updated from my previous version to v(3.4). Now, none of the versions are working, I tried removing and redownloading but the same error appears.

I would really appreciate any help with this issue.

Thank you in advance.

### PR DESCRIPTION
**Pull request by johnny-bit**
fixes #7592 - it however does not address the root cause. maybe it's an unchecked memory issue when allocation fails?

however it's better to have gtk yell at the console rather than outright crash

### CODE REVIEW

**Comment by dterrahe** [comment]
Any thoughts on setting G_DISABLE_CHECKS for the Release build? Hopefully conditions like these will be caught and fixed by developers using the RelWithDebInfo so there is a case to be made that for max speed they could be disabled. Any user reporting a crash without any further info from a Release build would then be asked to reproduce with RelWithDebInfo.

**Comment by johnny-bit** [comment]
Hmmm... TBH in order for bug from #7592 to happen, there has to be some very unlikely condition. Additionally the checks themselves are marked as "unlikely" so depending on available macros, compiler optimizations etc, predictors are told that the condition where widget is null is not likely to happen and if it were to happen... then we'd have problem anyway ;) So IMO the checks aren't intensive enough to clasify as "speed bumps" ;)

There's another issue, which requires some thinking: that proposition would require us to have "release" and "debug" packages available for users wishing to test. And nightlies that work ;) I'd build nightlies with debug info anyway...

**Comment by dterrahe** [comment]
I can't remember but thought the nightlies would be built with RelWithDebInfo?

I was just thinking of the person we know who always builds everything the fastest he can get, and he might not like having a lot of additional checking taking up time!

Hopefully it's clear that I'm somewhat joking here (gtk ui code is slow anyway, so an extra != NULL check isn't going to kill anyone) but in general I'm all for more checking and instrumenting, even in somewhat time critical code, as long as it can be switched off for release.

**Comment by parafin** [comment]
Disabling such checks is a good way to create undebuggable crashes, which disappear when you build less-optimized version (e.g. because race condition depends on optimization level) ;)

**Review by dterrahe — COMMENTED** [review]
> Disabling such checks is a good way to create undebuggable crashes, which disappear when you build less-optimized version (e.g. because race condition depends on optimization level) ;)

I guess as long as code keeps running because of side effects of debugging code, all is fine.

File: src/dtgtk/gradientslider.c [hunk 1]
```diff
@@ -166,6 +168,9 @@ static gint _get_active_marker_from_screen(GtkWidget *widget, const gdouble x, c
 
 static gdouble _slider_move(GtkWidget *widget, gint k, gdouble value, gint direction)
 {
+  g_return_val_if_fail(widget != NULL, value);
+  g_return_val_if_fail(DTGTK_IS_GRADIENT_SLIDER(widget), value);
```
[the remark this answers is not in the document]
Comment (johnny-bit) [reply src/dtgtk/gradientslider.c:171]: done :D

**Comment by dterrahe** [comment]
Now that you've officially asked me for a review...

I haven't dug into the details of all of the changes, sorry, but doesn't it look as if some of the static routine are directly called from the type implementation only? Doesn't that more or less guarantee that it is called for an instance of that type? Are these checks that gtk itself does too? (Would sort of be nice to be consistent with them, because it makes behavior more predictable for us).

_Personally_, but this is just a matter of taste, I would only check for NULLs if I'm dereferencing myself, because if I just pass it on to some gtk routines, then _those_ will already check and report a fail if there's nothing there. I always imagine stuff being passed up and down numerous chains and being checked again and again at every step. I also don't care too much about returning from a routine, after reporting a fail on the command line, rather than just chugging along and possibly crashing later. I'd rather see all the crashes in development rather than some developers chosing to ignore messages in the background and the result in production being wrong (where debugging might be harder).

I realise I'm not giving answers, just asking questions. There's nothing objectionable here, so it looks like it can safely be merged, but I'm not 100% convinced of the added value in all cases.

**Comment by johnny-bit** [comment]
The problem is - Without these we run into situation where we expect data/behaviour to be there while it isn't if we don't check. The things called might change over the time, maybe something gets called unnecesarily or with nul or whatever and it isn't too bad in and of itself, so we can 'let it slide' by just not doing anything and logging it. This gives us a graseful way to handle problem: either we have null + breackage somewhere in release, or in worst case user has log full of Gtk-CRITICAL errors which we should investigate :D 

I'd rather opt for full fix and keep check in place. Lose nothing in terms of speed (that's just non-null check) and gain no breackage :D

**Comment by johnny-bit** [comment]
I think this is mostly done. Debugging such issues could be done with `G_DEBUG=fatal-criticals` and running the app under debugger.

**Comment by Raza786000** [comment]
Thanks for your work on this problem. Is it available to download (3.4.1)?

**Comment by johnny-bit** [comment]
@Raza786000 - the PR isn't yet merged, however once merged I believe it'll become part of 3.4.1 release, therefore fixing your problem. (Well, not actually fixing root cause, but shouldn't crash, just yell a lot)

**Review by TurboGit — APPROVED** [review]
Looks good, thanks!

**Comment by TurboGit** [comment]
Merged in master and 3.4.1.

**Pull request merged**